src/signal_routing.rs

```diff
@@ -0,0 +1,235 @@
+/// Type of a gtk signal callback function that operates on uncast glib values.
+pub type RawSignalCallback = Box<dyn Fn(&[glib::Value]) -> Option<glib::Value>>;
+
+pub fn route_signal(
+    obj: impl glib::ObjectExt,
+    gtk_signal: &str,
+    actix_signal: &str,
+    target: impl IntoGenerateRoutingGtkHandler,
+) -> Result<glib::SignalHandlerId, crate::Error> {
+    let handler = target
+        .into_generate_routing_gtk_handler()
+        .generate_routing_gtk_handler(actix_signal);
+    let handler_id = obj.connect_local(gtk_signal, false, handler)?;
+    Ok(handler_id)
+}
+
+pub trait GenerateRoutingGtkHandler {
+    fn generate_routing_gtk_handler(&mut self, signal_name: &str) -> RawSignalCallback;
+}
+
+impl<T: Clone + 'static> GenerateRoutingGtkHandler for (T, actix::Recipient<crate::Signal<T>>) {
+    fn generate_routing_gtk_handler(&mut self, signal_name: &str) -> RawSignalCallback {
+        let signal_name = std::rc::Rc::new(signal_name.to_owned());
+        let (tag, recipient) = self.clone();
+        Box::new(move |parameters| {
+            if let Some(result) = crate::try_block_on(async {
+                let signal = crate::Signal::new(signal_name.clone(), parameters.to_owned(), tag.clone());
+                recipient.send(signal).await
+            }) {
+                let result = result.unwrap().unwrap();
+                if let Some(gtk::Inhibit(inhibit)) = result {
+                    use glib::value::ToValue;
+                    Some(inhibit.to_value())
+                } else {
+                    None
+                }
+            } else {
+                let signal = crate::Signal::new(signal_name.clone(), parameters.to_owned(), tag.clone());
+                recipient.do_send(signal).unwrap();
```

Comment:
Why does this always return `None` without handling the result value?

Reply:
I'm really unsure, but maybe what we want is `futures::executor::block_on(recipient.send(signal))`?

Reply:
We can't get the result of the message here - the message will only get processed _after_ we leave this function and exit the Actix context. `futures::executor::block_on` won't do - because we are working with Actix messages here, we can't just use any old `block_on` - we must use the `block_on` of the Actix system we are using. Note that if we could that would mean breaking the already lax concurrency rules Actix have, because if `try_block_on` returns `None` that means it gets executed inside the Actix runtime (our nested signals problem) when some actor is actually being borrowed mutably in order to handle an event - and this `block_on` could have send a message to it - which would either reborrow it or create a deadlock.

Reply:
But doesn't this code simply *swallow* results, potentially causing hard to spot logic bugs?

Reply:
Oh. Right. I did not care about the result's `Ok` value, but I neglected to consider its `Err` value...

I'll spawn a new future that unwraps these results. While I'm at it, it'll also make sure the signal indeed returned `None` - though if it didn't, GTK will probably fail earlier because it got `None` while expecting a boolean.

Reply:
Let us continue this discussion in #23.